### ISSUE TITLE
Win7/x64 SetFilePointerEx detour issue

### ISSUE THREAD

**Issue by randydu** [randydu/PolyHook_2_0#2]
The x64-detour algorithm does not patch *call* (FF15 xxxxxxxx) instruction correctly when it needs out-of-2G call target fix, as a result the trampoline code crashes on invalid call target.

## Before hooking

```
000000007739A3A0 | 48:83EC 28               | sub rsp,28                              |
000000007739A3A4 | E8 0B000000              | call <JMP.&SetFilePointerEx>            |
000000007739A3A9 | 48:83C4 28               | add rsp,28                              |
000000007739A3AD | C3                       | ret                                     |
000000007739A3AE | 90                       | nop                                     |
000000007739A3AF | 90                       | nop                                     |
000000007739A3B0 | 90                       | nop                                     |
000000007739A3B1 | 90                       | nop                                     |
000000007739A3B2 | 90                       | nop                                     |
000000007739A3B3 | 90                       | nop                                     |
000000007739A3B4 | FF25 76320900            | jmp qword ptr ds:[<&SetFilePointerEx>]  |
000000007739A3BA | 90                       | nop                                     |
000000007739A3BB | 90                       | nop                                     |
000000007739A3BC | 90                       | nop                                     |
```

## After hooking

```
000000007739A3A0 FF 25 F5 62 EC 88    jmp         qword ptr [26069Bh]  
000000007739A3A6 66 90                xchg        ax,ax  
000000007739A3A8 90                   nop  
000000007739A3A9 48 83 C4 28          add         rsp,28h  
000000007739A3AD C3                   ret  
000000007739A3AE 90                   nop  
000000007739A3AF 90                   nop  
```

Try running the JMP instruction throws exception:

```
Exception thrown at 0x000000007739A3A0 (kernel32.dll) in ut.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
```

What's inside the QWORD? 

```
0x000000000026069B  00 00 b8 04 01 00 00 4c 
```
Unfortunately, the correct target (detour fnCallback) should be: **0x000000013F6C5648**.

So it seems the searched code-cave is modified after hooking for some reason. This issue block us from reporting the FF15-patch issue mentioned above.

**Comment by randydu** [comment]
To further inspect the issue, launch another debugging session. This time, 

## After hook

```
000000007739A3A0 FF 25 F5 68 E3 88    jmp         qword ptr [1D0C9Bh]  
000000007739A3A6 66 90                xchg        ax,ax  
000000007739A3A8 90                   nop  
000000007739A3A9 48 83 C4 28          add         rsp,28h  
000000007739A3AD C3                   ret  
000000007739A3AE 90                   nop  
```

Trampoline code:
```
00000000001A3500 48 83 EC 28          sub         rsp,28h  
00000000001A3504 E8 AB 6E 1F 77       call        000000007739A3B4  
00000000001A3509 FF 25 46 00 00 00    jmp         qword ptr [1A3555h]  
00000000001A350F 90                   nop  
00000000001A3510 90                   nop  
00000000001A3511 90                   nop  
00000000001A3512 90                   nop  
00000000001A3513 90                   nop  
00000000001A3514 FF 25 76 32 09 00    jmp         qword ptr [236790h]  
00000000001A351A 90                   nop  
00000000001A351B 90                   nop  
00000000001A351C 90                   nop  
00000000001A351D 90                   nop  
00000000001A351E 90                   nop  
00000000001A351F 90                   nop  
00000000001A3520 EB 06                jmp         00000000001A3528  
00000000001A3522 90                   nop  
00000000001A3523 90                   nop  
00000000001A3524 90                   nop  
00000000001A3525 90                   nop  
00000000001A3526 90                   nop  
00000000001A3527 90                   nop  
00000000001A3528 FF 25 9A 31 09 00    jmp         qword ptr [2366C8h]  
00000000001A352E 90                   nop  
00000000001A352F 90                   nop  
00000000001A3530 90                   nop  
00000000001A3531 90                   nop  
00000000001A3532 90                   nop  
00000000001A3533 90                   nop  
00000000001A3534 90                   nop  
00000000001A3535 90                   nop  
00000000001A3536 90                   nop  
00000000001A3537 90                   nop  
00000000001A3538 90                   nop  
00000000001A3539 90                   nop  
00000000001A353A 90                   nop  
00000000001A353B 90                   nop  
00000000001A353C 90                   nop  
00000000001A353D 90                   nop  
00000000001A353E 90                   nop  
00000000001A353F 90                   nop  
00000000001A3540 4C 8B DC             mov         r11,rsp  
00000000001A3543 49 89 5B 10          mov         qword ptr [r11+10h],rbx  
00000000001A3547 49 89 73 18          mov         qword ptr [r11+18h],rsi  
00000000001A354B 49 89 4B 08          mov         qword ptr [r11+8],rcx  
00000000001A354F 57                   push        rdi  
00000000001A3550 41 54                push        r12  
00000000001A3552 41 55                push        r13  
00000000001A3554 41 A9 A3 39 77 00    test        eax,7739A3h  
00000000001A355A 00 00                add         byte ptr [rax],al  
00000000001A355C 00
```
We can see the original CALL (EB) target is within the 2G range and only need relocation to the call-target (000000007739A3B4 ). 

To inspect how the code-cave can be modified, I set a hardware data break on 1D0C9Bh immediately after the API is hooked.

Currently its content is _correct_ (the fnCallback):

```
0x00000000001D0C9B  48 56 7c 3f 01 00 00 00
```

When continue running the exe, the content of 0x!D0C9B is modified (but we do not catch any exception on the master machine, since the test exe is running in a remote Win7 machine).

After more testing, the result is random: sometimes it works fine, sometimes the fnCallback is modified, sometimes the content is _unmapped_ from process memory!

So I am pretty sure the current code-cave searching algorithm is not stable. The function  PLH::x64Detour::findNearestCodeCave(uint64_t addr) reads memory via (ReadProcessMemory) in the range of (-2G + api_addr, 2G + api_addr) and scans for all known code patterns until the first code-cave is found. It is not reliable because the process memory consists of many different types, such as heaps, stacks, virtual-allocated memory, file-mapping-views, etc.  the algorithm must make sure the code-cave is resolved from the stable memory area (not in any volatile memory range).

One solution is to search for module mapped memory for code-hole. we can enumerate the current mapped module list.

**Comment by randydu** [comment]
the workaround is using detour_scheme_t::INPLACE to hook this api.

**Comment by randydu** [comment]
Here is a proof that the code-cave algorithm can corrupt code:

In the following code, the code-cave pattern:

```c++ 
const char* NOP2_RET = "c3 0f 1f 44 00 00";
```

will corrupt the memset() code in memory:

Original Code:

memset.asm
```asm
   294: msetTab09:
   295:         mov     (-9)[rcx], rdx
00007FF70F3ACF28 48 89 51 F7          mov         qword ptr [rcx-9],rdx  
   296:         mov     (-1)[rcx], dl
00007FF70F3ACF2C 88 51 FF             mov         byte ptr [rcx-1],dl  
   297:         ret
00007FF70F3ACF2F C3                   ret  
   298: 
   299:         align   16
   300: msetTab13:
   301:         mov     (-13)[rcx], rdx
00007FF70F3ACF30 48 89 51 F3          mov         qword ptr [rcx-0Dh],rdx  
   302:         ; fallthrough to 5
   303: msetTab05:
   304:         mov     (-5)[rcx], edx
00007FF70F3ACF34 89 51 FB             mov         dword ptr [rcx-5],edx  
   305:         mov     (-1)[rcx], dl
00007FF70F3ACF37 88 51 FF             mov         byte ptr [rcx-1],dl  
   306:         ret
00007FF70F3ACF3A C3                   ret  
00007FF70F3ACF3B 0F 1F 44 00 00       nop         dword ptr [rax+rax]  
   307: 
   308:         align   16
   309: msetTab14:
   310:         mov     (-14)[rcx], rdx
00007FF70F3ACF40 48 89 51 F2          mov         qword ptr [rcx-0Eh],rdx  
   311:         ; fallthrough to 6
   312: msetTab06:
   313:         mov     (-6)[rcx], edx
00007FF70F3ACF44 89 51 FA             mov         dword ptr [rcx-6],edx  
   314:         ; fallthrough to 2
   315: msetTab02:
   316:         mov     (-2)[rcx], dx
00007FF70F3ACF47 66 89 51 FE          mov         word ptr [rcx-2],dx  
   317:         ret
00007FF70F3ACF4B C3                   ret  
   318: 
   319: msetTab08:
   320:         mov     [rax], rdx
00007FF70F3ACF4C 48 89 10             mov         qword ptr [rax],rdx  
   321:         ret
00007FF70F3ACF4F C3                   ret  
   322: 
   323: msetTab11:
   324:         mov     [rax], rdx
00007FF70F3ACF50 48 89 10             mov         qword ptr [rax],rdx  
   325:         mov     (8)[rax], dx
00007FF70F3ACF53 66 89 50 08          mov         word ptr [rax+8],dx  
   326:         mov     (10)[rax], dl
00007FF70F3ACF57 88 50 0A             mov         byte ptr [rax+0Ah],dl  
   327:         ret
00007FF70F3ACF5A C3                   ret  
00007FF70F3ACF5B 0F 1F 44 00 00       nop         dword ptr [rax+rax]  
   328: 
   329:         align   16
   330: msetTab10:
   331:         mov     [rax], rdx
00007FF70F3ACF60 48 89 10             mov         qword ptr [rax],rdx  
   332:         mov     (8)[rax], dx
00007FF70F3ACF63 66 89 50 08          mov         word ptr [rax+8],dx  
   333:         ret
00007FF70F3ACF67 C3                   ret  
   334: 
   335: msetTab16:
   336:         mov     [rax], rdx
00007FF70F3ACF68 48 89 10             mov         qword ptr [rax],rdx  
   337:         mov     (8)[rax], rdx
00007FF70F3ACF6B 48 89 50 08          mov         qword ptr [rax+8],rdx  
   338:         ret
00007FF70F3ACF6F C3                   ret  
```

The corrupted code:
```asm
   294: msetTab09:
   295:         mov     (-9)[rcx], rdx
00007FF70F3ACF28 48 89 51 F7          mov         qword ptr [rcx-9],rdx  
   296:         mov     (-1)[rcx], dl
00007FF70F3ACF2C 88 51 FF             mov         byte ptr [rcx-1],dl  
   297:         ret
00007FF70F3ACF2F C3                   ret  
00007FF70F3ACF30 48                   ?? ?????? 
00007FF70F3ACF31 89                   ?? ?????? 
00007FF70F3ACF32 51                   ?? ?????? 
00007FF70F3ACF33 F3                   ?? ?????? 
00007FF70F3ACF34 89                   ?? ?????? 
00007FF70F3ACF35 51                   ?? ?????? 
00007FF70F3ACF36 FB                   ?? ?????? 
00007FF70F3ACF37 88                   ?? ?????? 
00007FF70F3ACF38 51                   ?? ?????? 
00007FF70F3ACF39 FF                   ?? ?????? 
00007FF70F3ACF3A C3                   ?? ?????? 
00007FF70F3ACF3B 00                   ?? ?????? 
00007FF70F3ACF3C 00                   ?? ?????? 
00007FF70F3ACF3D D8                   ?? ?????? 
00007FF70F3ACF3E 9A                   ?? ?????? 
00007FF70F3ACF3F 5F                   ?? ?????? 
00007FF70F3ACF40 02                   ?? ?????? 
00007FF70F3ACF41 00                   ?? ?????? 
00007FF70F3ACF42 00                   ?? ?????? 
00007FF70F3ACF43 F2                   ?? ?????? 
00007FF70F3ACF44 89                   ?? ?????? 
00007FF70F3ACF45 51                   ?? ?????? 
00007FF70F3ACF46 FA                   ?? ?????? 
00007FF70F3ACF47 66                   ?? ?????? 
00007FF70F3ACF48 89                   ?? ?????? 
00007FF70F3ACF49 51                   ?? ?????? 
00007FF70F3ACF4A FE                   ?? ?????? 
00007FF70F3ACF4B C3                   ?? ?????? 
00007FF70F3ACF4C 48                   ?? ?????? 
00007FF70F3ACF4D 89                   ?? ?????? 
00007FF70F3ACF4E 10                   ?? ?????? 
00007FF70F3ACF4F C3                   ?? ?????? 
00007FF70F3ACF50 48                   ?? ?????? 
00007FF70F3ACF51 89                   ?? ?????? 
00007FF70F3ACF52 10                   ?? ?????? 
00007FF70F3ACF53 66                   ?? ?????? 
00007FF70F3ACF54 89                   ?? ?????? 
00007FF70F3ACF55 50                   ?? ?????? 
00007FF70F3ACF56 08                   ?? ?????? 
00007FF70F3ACF57 88                   ?? ?????? 
00007FF70F3ACF58 50                   ?? ?????? 
00007FF70F3ACF59 0A                   ?? ?????? 
00007FF70F3ACF5A C3                   ?? ?????? 
00007FF70F3ACF5B 00                   ?? ?????? 
00007FF70F3ACF5C 00                   ?? ?????? 
00007FF70F3ACF5D D8                   ?? ?????? 
00007FF70F3ACF5E 9A                   ?? ?????? 
--- d:\A01\_work\12\s\src\vctools\crt\vcruntime\src\string\amd64\memset.asm ----
   327:         ret
00007FF70F3ACF5F 5F                   pop         rdi  
   328: 
   329:         align   16
   330: msetTab10:
   331:         mov     [rax], rdx
00007FF70F3ACF60 02 00                add         al,byte ptr [rax]  
00007FF70F3ACF62 00 66 89             add         byte ptr [rsi-77h],ah  
   332:         mov     (8)[rax], dx
00007FF70F3ACF65 50                   push        rax  
00007FF70F3ACF66 08 C3                or          bl,al  
   334: 
   335: msetTab16:
   336:         mov     [rax], rdx
00007FF70F3ACF68 48 89 10             mov         qword ptr [rax],rdx  
   337:         mov     (8)[rax], rdx
00007FF70F3ACF6B 48 89 50 08          mov         qword ptr [rax+8],rdx  
   338:         ret
```

**Comment by randydu** [comment]
The code-cave too small bug is fixed in commit: https://github.com/randydu/PolyHook_2_0/commit/81786818cd8ead61a0103091d193eb8e81e3b72e

**Comment by stevemk14ebr** [comment]
oh darn yea there's a more general issue in that i only check SIZE against the amount we were able to read rather than actually inspecting if that read memory is pad bytes or code or not, hmm. This fix you've done so far is fine, but I need to do that more general fix of validating cave size. 

I really hate the code cave finder. I had a really hard time figuring out where to shove the extra bytes for the jump destination. I originally had a fancy memory allocator that could guarantee allocations within some address range but that failed when there was too many hook or on a constrained system. So then i thought i'd just find cave's and use those but then there's issue with the cave life time like you've noticed as well. I'm really not sure how to do this part the best, it's probably the single biggest issue with x64 hooks

**Comment by randydu** [comment]
> oh darn yea there's a more general issue in that i only check SIZE against the amount we were able to read rather than actually inspecting if that read memory is pad bytes or code or not, hmm. This fix you've done so far is fine, but I need to do that more general fix of validating cave size.
> 
Yes, adds a cave-size parameter to PLH::x64Detour::findNearestCodeCave(uint64_t addr) so only big-enough-cave can be recycled ;-)  

> I really hate the code cave finder. I had a really hard time figuring out where to shove the extra bytes for the jump destination. I originally had a fancy memory allocator that could guarantee allocations within some address range but that failed when there was too many hook or on a constrained system. So then i thought i'd just find cave's and use those but then there's issue with the cave life time like you've noticed as well. I'm really not sure how to do this part the best, it's probably the single biggest issue with x64 hooks

What about query loaded module list via PEB? for each module, if the exe/dll's PE header is available (most likely), search for padded bytes between PE sections, usually there are lots of wasted bytes between loaded PE sections  (Section_k.VirtualAddress + Section_k.VirtualSize, Section_k+1.VirtualAddress). Also, you can inspect PE image's resources, there are many bytes wasted due to section alignment. If lucky, many PE files have *Debug* directory (even for release version), which is useless at runtime, the debug data can be recycled; the old DOS-stub in PE header is also useless and can be reused. Of course, only modules mapped within the (addr-2G, addr+2G) should be scanned for code-cave.

**Comment by stevemk14ebr** [comment]
I would merge a solution that uses the code cave finder logic and applies it first the sections of the modules in the PEB like you suggest, then falls back to the terrible raw scan I do now as a last ditch effort. I will probably not find time to do this myself, but that sounds like a very good solution for the common case at least

**Comment by stevemk14ebr** [comment]
Hi, I have a solution but it restricts the usage of the library to windows 10 or whichever platforms implement VirtualAlloc2: https://twitter.com/stevemk14ebr/status/1389951007770091521?s=20

**Comment by randydu** [comment]
> Hi, I have a solution but it restricts the usage of the library to windows 10 or whichever platforms implement VirtualAlloc2: https://twitter.com/stevemk14ebr/status/1389951007770091521?s=20

Cool, we can first GetProcAddress() to detect if VirtaulAlloc2 api is available and try allocating the memory, if the memory allocation fails, fall back to other tricks.

**Comment by stevemk14ebr** [comment]
Implemented this here: https://github.com/stevemk14ebr/PolyHook_2_0/commit/73f38e8b7792ebee60043dd45c39962c5a004b13

**Comment by randydu** [comment]
fixed in v0.2.0 after all small code-caves are excluded.

Closed.

**Comment by stevemk14ebr** [comment]
I have pushed a better implementation just now. This new work is a replacement for the old PageAllocator. The new RangeAllocator uses VirtualAlloc2 if it is supported, then falls back to the old VirtualQuery walk logic but i've fixed bugs now that make that work way more reliably! The new system also has a true allocator on top of it that will split up the large chunks returned by virtualalloc. I'm very happy with the system now, it should be significantly more reliable.
